pkg/nsx/services/vpc/vpc.go

```diff
@@ -471,3 +514,230 @@ func (s *VPCService) Cleanup() error {
 
 	return nil
 }
+
+func (service *VPCService) needUpdateRule(rule *model.Rule, externalCIDR []string) bool {
+	des := rule.DestinationGroups
+	currentDesSet := sets.Set[string]{}
+	for _, group := range des {
+		currentDesSet.Insert(group)
+	}
+	if len(externalCIDR) != len(currentDesSet) {
+		return true
+	}
+	for _, newGroup := range externalCIDR {
+		if !currentDesSet.Has(newGroup) {
+			return true
+		}
+	}
+	return false
+}
+
+func (service *VPCService) getExternalIpBlock(vpcName string, namespace string) ([]string, error) {
+	// read corresponding vpc network config from store
+	nc_name, err := service.getNetworkconfigNameFromNS(namespace)
+	if err != nil {
+		log.Error(err, "failed to get network config name for VPC when creating NSX VPC rule", "VPC", vpcName, "namespace", namespace)
+		return nil, err
+	}
+	nc, _exist := service.GetVPCNetworkConfig(nc_name)
+	if !_exist {
+		log.Error(err, "failed to read network config  when creating NSX VPC rule", "nc_name", nc_name)
+		return nil, err
+	}
+	log.Info("get external ip block", "ip block", nc.ExternalIPv4Blocks)
+	return nc.ExternalIPv4Blocks, nil
+
+}
+
+func (service *VPCService) getIpblockCidr(cidrID []string) (result []string, err error) {
+	for _, cidr := range cidrID {
+		ipblock := service.PubIpblockStore.GetByKey(cidr)
+		if ipblock == nil {
+			err = errors.New("ipblock not found")
+			log.Error(err, "failed to get public ipblock", "path", cidr)
+			return
+		} else {
+			result = append(result, *ipblock.Cidr)
+		}
+	}
+	return
+}
+
+func (service *VPCService) CreateOrUpdateAVISecurityPolicy(vpc *model.Vpc, namespace string) error {
+	if !service.supportAviAllowRule {
+		return nil
+	}
+	path := strings.Split(*vpc.Path, "/")
+	orgId := path[2]
+	projectId := path[4]
+	ruleId := "allow-all-avi-se"
+	groupId := "all-avi-se-vms"
+	spId := "default-layer3-section"
+
+	if !service.existAVISecurityPolicy(orgId, projectId, *vpc.Id, spId) {
+		return errors.New("avi security policy not found")
+	}
+	allowrule, err := service.getAVIAllowRule(orgId, projectId, *vpc.Id, spId, ruleId)
+	var externalCIDR []string
+	if err == nil {
+		externalCIDR, err = service.getIpblockCidr(vpc.ExternalIpv4Blocks)
+		if err != nil {
+			return err
+		}
+	}
+
+	if allowrule != nil {
+		if !service.needUpdateRule(allowrule, externalCIDR) {
+			log.Info("avi rule no changed, skip update aiv rule")
+			return nil
+		} else {
+			log.Info("avi rule changed", "previous", allowrule.DestinationGroups, "current", externalCIDR)
+		}
+	}
+
+	group, err := service.getAVIGroup(orgId, projectId, *vpc.Id, groupId)
+	if err != nil {
+		log.Info("create avi group", "group", group)
+		group, err = service.createAVIGroup(orgId, projectId, *vpc.Id, groupId)
+		if err != nil {
+			log.Error(err, "failed to create avi group", "group", groupId)
```

Comment:
could we combine get and create AVI group in one func?

Reply:
add a new function getorCreateAVIGroup